### PR TITLE
Old transaction after resend was not removed from pool

### PR DESCRIPTION
**Pull request by bas-vk**
After [this change](https://github.com/ethereum/go-ethereum/pull/1241/files#diff-5ca161f6ea16137a30ef1967569c8418R966) the transaction hash stored in the transaction pool is of the signed transaction. This causes the `eth.resend` transaction not to remove the pending transaction when it was resend since the removal is on the unsigned transaction returned by `eth.pendingTransactions`.

Part of the signature is a random nonce which makes it impossible to recreate the signed transaction and remove the transaction using the old method. Therefore the transaction Hash was added to `eth.pendingTransactions` and `txPool.RemoveTx` function was exported which accepts this hash and removed the transaction from the pool.

### CODE REVIEW

File: rpc/api/eth_args.go [hunk 1]
```diff
@@ -931,6 +933,12 @@ func (tx *tx) UnmarshalJSON(b []byte) (err error) {
 		contractCreation = true
 	)
 
+	if val, found := fields["Hash"]; found {
```
Comment (Gustav-Simonsson) [inline rpc/api/eth_args.go]: Does not setting the hash from API input open up for errors, e.g. caller sending the wrong hash and it would be set as the tx hash. Seems better if we calculate it for the caller, or verifying the hash they provide.

**Comment by bas-vk** [comment]
After @Gustav-Simonsson comment an additional check is added which verifies if the user has supplied the correct, nonce, gas price + limit, recipient, value and data for the transaction to be resend.

File: rpc/api/eth.go [hunk 2]
```diff
@@ -578,14 +579,15 @@ func (self *ethApi) Resend(req *shared.Request) (interface{}, error) {
 		return nil, shared.NewDecodeParamError(err.Error())
 	}
 
-	ret, err := self.xeth.Transact(args.Tx.From, args.Tx.To, args.Tx.Nonce, args.Tx.Value, args.GasLimit, args.GasPrice, args.Tx.Data)
-	if err != nil {
-		return nil, err
+	pending := self.ethereum.TxPool().GetTransactions()
+	for _, p := range pending {
+		if found := p.SigHash() == args.Tx.tx.SigHash(); found {
```
Comment (fjl) [inline rpc/api/eth.go]: `SigHash` is expensive. Maybe that doesn't matter in this case, just informing you about it.
Might be better to compare `From` and `Nonce`. They uniquely identify the transaction for the purpose of resend.

File: core/transaction_pool.go [hunk 3]
```diff
@@ -375,6 +376,7 @@ func (pool *TxPool) removeTx(hash common.Hash) {
 			break
 		}
 	}
+	return
```
Comment (fjl) [inline core/transaction_pool.go]: Why was this added?

File: core/transaction_pool.go [hunk 4]
```diff
@@ -356,11 +356,12 @@ func (self *TxPool) RemoveTransactions(txs types.Transactions) {
...
 	defer self.mu.Unlock()
 	for _, tx := range txs {
-		self.removeTx(tx.Hash())
+		self.RemoveTx(tx.Hash())
 	}
 }
 
-func (pool *TxPool) removeTx(hash common.Hash) {
+// Remove transaction from the pool by its hash
+func (pool *TxPool) RemoveTx(hash common.Hash) {
```
Comment (fjl) [inline core/transaction_pool.go]: Note: conventional form for documentation format is
```go
// RemoveTx removes the transaction with the given hash from the pool.
```

(starts with function name, full sentence).

**Comment by bas-vk** [comment]
@fjl, I don't expect that this method will cause performance issues since it's not part of the official RPC interface and developers cannot rely on its presence. So it this method should only be called during testing/debugging.

I have added a check on the `From` address. The `Nonce` is part of the `SigHash` and part of the comparison. Or am I misunderstanding you?

The docs have been upgraded according to the convention and the unnecessary return is removed.

**Comment by obscuren** [comment]
:+1:

**Comment by fjl** [comment]
`SigHash` is expensive because it serializes the tx and the hash is not cached. This can be an issue if there are many pending transactions so I'd rather avoid using it. Comparing `From` and `Nonce` should really be fine and is much cheaper.

Another question: shouldn't the txpool replace the transaction if it receives a different one with the same nonce? I think it should, otherwise resends will not work on the network. Or maybe I'm not getting it.

**Comment by obscuren** [comment]
@fjl no the transaction pool should not replace a transaction with the same nonce and address. This helps the free market and flow of transactions over the network and allows miners to pick whatever they prefer.

**Comment by fjl** [comment]
So it should keep both?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          ‎

**Comment by obscuren** [comment]
No. Resend (name is debatable) deletes old and sends new (or resents equal hash). Creating a handcrafted tx (by setting the nonce) does not delete since it goes directly in the to pool.

**Comment by fjl** [comment]
Thanks. Use of `SigHash` is fine then. :+1:

**Comment by obscuren** [comment]
I believe the `resend` message should change name.
